File: src/scc.ts
const FREQ_BITS = 16;
const DEF_GAIN = 8;

class Channel {
    key: number;
    volume: number;
    counter: number;
    frequency: number;
    waveram: Array<number>;

    constructor() {
        this.key = 0;
        this.volume = 0;
        this.counter = 0;
        this.frequency = 0;
        this.waveram = new Array<number>(32);
    }
}

export default class SCC {
    voice: Array<Channel>;
    rate: number;
    clock: number;

    output: Float64Array;

    mixerTable: Int16Array;
    mixerLookup: number;

    test: number;
    currentRegister: number;

    process(samples: number): void {
        const mixerBuffer = new Int16Array(this.output.length);

        for (let channel = 0; channel < 5; channel++) {
            const frequency = this.voice[channel].frequency;

            // Channel is halted for frequency < 9
            if (frequency > 8) {
                const waveram = this.voice[channel].waveram;
                const volume = this.voice[channel].volume * this.voice[channel].key;
                let counter = this.voice[channel].counter;

                const step = ((this.clock * (1 << FREQ_BITS)) / ((frequency + 1) * 16 * (this.rate / 32)) + 0.5);

                for (let i = 0; i < samples; i++) {
                    counter += step;
                    const offset = (counter >> FREQ_BITS) & 0x1F;
                    mixerBuffer[i] += (waveram[offset] * volume) >> 3;
                }

                this.voice[channel].counter = counter;
            }
        }

        for (let i = 0; i < samples; i++) {
            mixerBuffer[i] = this.mixerTable[this.mixerLookup + mixerBuffer[i]];
            this.output[i] = mixerBuffer[i] / 32768;
        }
    }

    setWaveform(offset: number, data: number): void {
        // Is RAM read-only?
        if (this.test & 0x40 || (this.test & 0x80 && offset >= 0x60)) return;

        if (offset >= 0x60) {
            // Channel 5 shares RAM with channel 4
            this.voice[3].waveram[offset & 0x1F] = data;
            this.voice[4].waveram[offset & 0x1F] = data;
        } else this.voice[offset >> 5].waveram[offset & 0x1F] = data;
    }

    getWaveform(offset: number): number {
        // Test register bits 6/7 expose the internal counter
        if (this.test & 0xC0) {
            if (offset >= 0x60)
                offset += (this.voice[3 + (this.test >> 6 & 1)].counter >> FREQ_BITS);
            else if (this.test & 0x40)
                offset += (this.voice[offset >> 5].counter >> FREQ_BITS);
        }

        return this.voice[offset >> 5].waveram[offset & 0x1F];
    }

    setWaveformPlus(offset: number, data: number): void {
        // Is RAM read-only?
        if (this.test & 0x40) return;

        this.voice[offset >> 5].waveram[offset & 0x1F] = data;
    }

    getWaveformPlus(offset: number): number {
        // Test register bit 6 exposes the internal counter
        if (this.test & 0x40)
            offset += (this.voice[offset >> 5].counter >> FREQ_BITS);

        return this.voice[offset >> 5].waveram[offset & 0x1F];
    }

    setVolume(offset: number, data: number): void {
        this.voice[offset & 0x7].volume = data & 0xF;
    }

    setFrequency(offset: number, data: number): void {
        const channel = this.voice[offset >> 1];

        // Test register bit 5 resets the internal counter
        if (this.test & 0x20)
            channel.counter = ~0;
        else if (channel.frequency < 9)
            channel.counter |= ((1 << FREQ_BITS) - 1);

        if (offset & 1)
            channel.frequency = (channel.frequency & 0x0FF) | ((data << 8) & 0xF00);
        else
            channel.frequency = (channel.frequency & 0xF00) | (data << 0);

        // Valley Bell: Behaviour according to OpenMSX
        channel.counter &= 0xFFFF0000;
    }

    setKeyOnOff(data: number): void {
        for (let channel = 0; channel < 5; channel++) {
            this.voice[channel].key = data & 1;
            data >>= 1;
        }
    }

    setTest(data: number): void {
        this.test = data;
    }

    getTest(): number {
        // Reading the test register sets it to 0xFF
        this.setTest(0xFF);
        return 0xFF;
    }

    writeRegister(offset: number, data: number): void {
        switch (offset & 1) {
            case 0x00:
                this.currentRegister = data;
                break;
            case 0x01:
                switch (offset >> 1) {
                    case 0x00:
                        this.setWaveform(this.currentRegister, data);
                        break;
                    case 0x01:
                        this.setFrequency(this.currentRegister, data);
                        break;
                    case 0x02:
                        this.setVolume(this.currentRegister, data);
                        break;
                    case 0x03:
                        this.setKeyOnOff(data);
                        break;
                    case 0x04:
                        this.setWaveformPlus(this.currentRegister, data);
                        break;
                    case 0x05:
                        this.setTest(data);
                        break;
                }
                break;
        }
    }

    constructor(clock: number, sampleRate: number) {
        this.voice = [
            new Channel(),
            new Channel(),
            new Channel(),
            new Channel(),
            new Channel()
        ];

        this.clock = clock & 0x7FFFFFFF;
        this.rate = this.clock / 16;

        this.output = new Float64Array(sampleRate);

        this.mixerTable = new Int16Array(2561);
        this.mixerLookup = 1280;

        for (let i = 0; i < 1281; i++) {
            let val = i * DEF_GAIN * 16 / 5;
            if (val > 32768) val = 32768;
            this.mixerTable[this.mixerLookup + i] = val;
            this.mixerTable[this.mixerLookup - i] = -val;
        }

        this.currentRegister = 0;
        this.test = 0;
    }
}

File: src/scc.zig
const std = @import("std");
const math = std.math;
const panic = std.debug.panic;
const print = std.debug.warn;

const VOICES = 5;
const FREQ_BITS = 16;
const DEF_GAIN = 8;
const AMP_FACTOR = 256;
const AMP_LOOKUP = VOICES * AMP_FACTOR;
const AMP_SIZE = (AMP_LOOKUP * 2) + 1;

pub const Chip = struct {
    voice: [VOICES]Channel,
    amp_table: [AMP_SIZE]i16,
    output: f64,
    const Self = @This();

    pub fn init(clock: u32, sample_rate: u32) Self {
        var chip = Chip{
            .voice = [_]Channel{Channel{
                .period = 0,
                .counter = 0,
                .waveform = [_]i8{0} ** 32,
                .volume = 0,
                .mute = 0,
                .output = 0,
                .clock = 0,
                .rate = 0,
            }} ** VOICES,
            .amp_table = [_]i16{0} ** AMP_SIZE,
            .output = 0.0,
        };

        var channel: usize = 0;
        const channel_clock = clock & 0x7FFFFFFF;
        const channel_rate = channel_clock / 16;
        while (channel < VOICES) : (channel += 1) {
            chip.voice[channel].clock = channel_clock;
            chip.voice[channel].rate = channel_rate;
        }

        var i: u32 = 0;
        while (i < AMP_LOOKUP) : (i += 1) {
            var value = (AMP_LOOKUP - i) * DEF_GAIN * 16 / VOICES;
            if (value > 32767) value = 32767;
            const negative_value = math.negateCast(@intCast(i16, value)) catch |err| panic("ERROR: {}", .{err});
            chip.amp_table[i] = negative_value;
        }
        i = 0;
        while (i < (AMP_LOOKUP + 1)) : (i += 1) {
            var value = i * DEF_GAIN * 16 / VOICES;
            if (value > 32767) value = 32767;
            chip.amp_table[AMP_LOOKUP + i] = @intCast(i16, value);
        }

        return chip;
    }

    pub fn writeRegister(s: *Self, port: u8, register: u8, data: u8) void {
        switch (port) {
            // SCC mode waveform write
            0 => {
                if (port >= 0x60) {
                    // Channel 4 and 5 share RAM
                    s.voice[3].setWaveform(register, data);
                    s.voice[4].setWaveform(register, data);
                } else {
                    s.voice[register >> 5].setWaveform(register, data);
                }
            },
            1 => s.voice[register >> 1].setFrequency(register, data),
            2 => s.voice[register & 0x7].setVolume(data),
            3 => {
                var d = data;
                var channel: u3 = 0;
                while (channel < VOICES) : (channel += 1) {
                    s.voice[channel].setMute(d);
                    d >>= 1;
                }
            },
            // SCC-I mode waveform write
            4 => s.voice[register >> 5].setWaveform(register, data),
            5 => {}, // Test register not implemented
            else => unreachable,
        }
    }

    pub fn process(s: *Self) void {
        var mixer: i32 = 0;
        var channel: usize = 0;

        while (channel < VOICES) : (channel += 1) {
            s.voice[channel].update();
            mixer += s.voice[channel].output;
        }

        const tone = s.amp_table[@intCast(usize, AMP_LOOKUP + mixer)];
        s.output = @intToFloat(f64, tone);
    }
};

const Channel = struct {
    period: u12,
    counter: u64,
    waveform: [32]i8,
    volume: u4,
    mute: u1,
    output: i32,
    clock: u32,
    rate: u32,
    const Self = @This();

    fn setWaveform(s: *Self, register: u8, data: u8) void {
        const address = @intCast(usize, register & 0x1F);
        s.waveform[address] = @intCast(i8, data);
    }

    fn setFrequency(s: *Self, register: u8, data: u8) void {
        const d = @intCast(u12, data);
        if ((register & 1) != 0) {
            s.period = (s.period & 0x0FF) | ((d << 8) & 0xF00);
        } else {
            s.period = (s.period & 0xF00) | (d << 0);
        }
        s.counter &= 0xFFFF0000; // Behaviour according to openMSX
    }

    fn setVolume(s: *Self, data: u8) void {
        s.volume = @intCast(u4, data & 0xF);
    }

    fn setMute(s: *Self, data: u8) void {
        s.mute = @intCast(u1, data & 1);
    }

    fn update(s: *Self) void {
        if (s.period < 9) return; // Channel is halted for period < 9

        const volume = s.volume * s.mute;

        // Amuse source:  Cab suggests this method gives greater resolution
        // Sean Young 20010417: the formula is really: f = clock/(16*(f+1))
        const clock = @intCast(i64, s.clock);
        const period = @intToFloat(f64, s.period);
        const rate = @intToFloat(f64, s.rate);

        const clock_shifted = @intToFloat(f64, clock * (1 << FREQ_BITS));
        const rate_shifted = ((period + 1) * 16 * (rate / 32) + 0.5);
        const step = @floatToInt(u32, clock_shifted / rate_shifted);

        s.counter += step;
        const address = (s.counter >> FREQ_BITS) & 0x1F;
        s.output = (s.waveform[address] * @intCast(i32, volume)) >> 3;
    }
};
